src/Models/boundary_adjacent_mean.jl

```diff
@@ -0,0 +1,156 @@
+├── east: OpenBoundaryCondition{Oceananigans.BoundaryConditions.PerturbationAdvection{Val{true}, Float64}}: BoundaryAdjacentMean: (-1.5612511283791264e-18)
+├── south: DefaultBoundaryCondition (FluxBoundaryCondition: Nothing)
+├── north: DefaultBoundaryCondition (FluxBoundaryCondition: Nothing)
+├── bottom: DefaultBoundaryCondition (FluxBoundaryCondition: Nothing)
+├── top: DefaultBoundaryCondition (FluxBoundaryCondition: Nothing)
+└── immersed: DefaultBoundaryCondition (FluxBoundaryCondition: Nothing)
+
+```
+"""
+struct BoundaryAdjacentMean{NI, BV}
+  boundary_parallel_integral :: NI
+                       value :: BV
+
+   BoundaryAdjacentMean(grid, side; 
+                        boundary_parallel_integral::NI = boundary_reduced_field(Val(side), grid),
+                        value::BV = Ref(zero(grid))) where {NI, BV} = 
+        new{NI, BV}(boundary_parallel_integral, value)
+end
+
+@inline (bam::BoundaryAdjacentMean)(args...) = bam.value[]
+
+Adapt.adapt_structure(to, mo::BoundaryAdjacentMean) = 
+    BoundaryAdjacentMean(; boundary_parallel_integral = nothing, value = adapt(to, mo.value[]))
+
+Base.show(io::IO, bam::BoundaryAdjacentMean) = print(io, summary(bam)*"\n")
+Base.summary(bam::BoundaryAdjacentMean) = "BoundaryAdjacentMean: ($(bam.value[]))"
+
+@inline boundary_reduced_field(::Union{Val{:west}, Val{:east}}, grid)   = Field{Face, Nothing, Nothing}(grid)
+@inline boundary_reduced_field(::Union{Val{:south}, Val{:north}}, grid) = Field{Nothing, Face, Nothing}(grid)
+@inline boundary_reduced_field(::Union{Val{:bottom}, Val{:top}}, grid)  = Field{Nothing, Nothing, Face}(grid)
+
+@inline boundary_normal_area(::Union{Val{:west}, Val{:east}}, grid)   = GridMetricOperation((Face, Center, Center), Ax, grid)
+@inline boundary_normal_area(::Union{Val{:south}, Val{:north}}, grid) = GridMetricOperation((Center, Face, Center), Ay, grid)
+@inline boundary_normal_area(::Union{Val{:bottom}, Val{:top}}, grid)  = GridMetricOperation((Center, Center, Face), Az, grid)
+
+@inline boundary_adjacent_indices(::Val{:east}, grid, loc) = size(grid, 1)+1, 1, 1
+@inline boundary_adjacent_indices(val_side::Val{:west}, grid, loc) = 1, 1, 1
+
+@inline boundary_adjacent_indices(::Val{:north}, grid, loc) = 1, size(grid, 2)+1, 1
+@inline boundary_adjacent_indices(val_side::Val{:south}, grid, loc) = 1, 1, 1
+
+@inline boundary_adjacent_indices(::Val{:top}, grid, loc) = 1, 1, size(grid, 3)+1
+@inline boundary_adjacent_indices(val_side::Val{:bottom}, grid, loc) = 1, 1, 1
+
+(bam::BoundaryAdjacentMean)(side, u) = bam(Val(side), u)
+
+# computes the boundary mean and stores/returns it
+function (bam::BoundaryAdjacentMean)(val_side::Val, u)
+    grid = u.grid
+
+    loc = instantiated_location(u)
+    iB, jB, kB = boundary_adjacent_indices(val_side, grid, loc)
+    An = boundary_normal_area(val_side, grid)
+
+    # get the total flux
+    sum!(bam.boundary_parallel_integral, (@at (Face, Center, Center) u) * An)
```

Comment:
A few comments:

1. It might be nicer to preform this operation --- that way users can look at it, and potentially verify its correctness.
2. Why don't we use `Average` here?
3. `sum!` is incorrect if the grid is stretched, right?

Reply:
Do you mean to return the operation and let the user do it?

I can't exactly remember how I ended up with this design but I think I did it this was so that we don't have to allocate a new average field (since you can't do in place averaging right?)

And I think sum is correct since it is multiplied by the area right?